src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs

```diff
@@ -90,233 +95,109 @@ public void Dispose()
-                else if (extractedChild is TForeachStatementSyntax foreachStatement &&
-                    TryProcessForeachStatement(foreachStatement))
-                {
-                    continue;
-                }
-                else if (extractedChild is TIfStatementSyntax ifStatement &&
-                    TryProcessIfStatement(ifStatement))
-                {
-                    continue;
-                }
+                cancellationToken.ThrowIfCancellationRequested();
+                var match = TryAnalyzeStatement(statement, ref seenInvocation, ref seenIndexAssignment, cancellationToken);
+                if (match is null)
+                    break;
 
-                // Something we didn't understand. Stop here.
-                break;
+                matches.Add(match.Value);
             }
 
             return true;
+        }
 
-            bool TryProcessExpressionStatement(TExpressionStatementSyntax expressionStatement)
-            {
-                // Can't mix Adds and indexing.
-                if (!seenIndexAssignment)
-                {
-                    // Look for a call to Add or AddRange
-                    if (TryAnalyzeInvocation(
-                            expressionStatement,
-                            addName: WellKnownMemberNames.CollectionInitializerAddMethodName,
-                            requiredArgumentName: null,
-                            out var instance) &&
-                        ValuePatternMatches(instance))
-                    {
-                        seenInvocation = true;
-                        matches.Add(new Match<TStatementSyntax>(expressionStatement, UseSpread: false));
-                        return true;
-                    }
-                    else if (
-                        _analyzeForCollectionExpression &&
-                        TryAnalyzeInvocation(
-                            expressionStatement,
-                            addName: nameof(List<int>.AddRange),
-                            requiredArgumentName: null,
-                            out instance))
-                    {
-                        seenInvocation = true;
-
-                        // AddRange(x) will become `..x` when we make it into a collection expression.
-                        matches.Add(new Match<TStatementSyntax>(expressionStatement, UseSpread: true));
-                        return true;
-                    }
-                }
+        private Match<TStatementSyntax>? TryAnalyzeStatement(
+            TStatementSyntax statement, ref bool seenInvocation, ref bool seenIndexAssignment, CancellationToken cancellationToken)
+        {
+            return _analyzeForCollectionExpression
+                ? State.TryAnalyzeStatementForCollectionExpression(this.SyntaxHelper, statement, cancellationToken)
+                : TryAnalyzeStatementForCollectionInitializer(statement, ref seenInvocation, ref seenIndexAssignment, cancellationToken);
+        }
```

Comment:
the general idea is:

1. analysis for a collection *initializer* only ever needs to be done here.  so the logic for that is here.
2. analysis for a collection-expression needs to be done here, and also in teh collection-builder case, so it moved to a common location that both of them can use.  

This made the logic in this type also simpler as it only has to handle teh initializer case. so no need to deal with foreach/if-statements, or look for AddRange statements since collection-initializers can't handle those.